Extract database specific adapters.

sahilm opened the pull request:
- This is to support skipping backup and restore
  on SQLServer databases.

- We want to skip all steps in backup and restore when
  deploying to a database on which we don't support such
  tasks yet.

sahilm commented:
Tested against PostgreSQL, MySQL and SQL Server.

vipulnsward commented:
Will verify this tomorrow.

vipulnsward commented on lib/boxy-cap/database_adapter.rb:
```diff
@@ -0,0 +1,86 @@
+require 'rake'
+
+class DatabaseAdapter
```
Lets move to `database_adapters` directory and call it 'BaseAdapter'

sahilm replied:
* If I call it `BaseAdapter` the api will change from `DatabaseAdapter.for` to `BaseAdapter.for`. It will not be clear what the `BaseAdapter` does from the api.

* I will move it into `database_adapters`.

vipulnsward replied:
Just call it `Base` maybe. Method name can be bigger. Also all these should be namespace `DatabaseAdapters` since they lie in that directory. In that case call would be `DatabaseAdapters::Base.adapter_for` or something similar.

sahilm replied:
Alright. Will do.

On Thu, Jun 18, 2015 at 8:14 PM, Vipul A M <notifications@github.com> wrote:

> In lib/boxy-cap/database_adapter.rb
> <https://github.com/bigbinary/boxy-cap/pull/2#discussion_r32737430>:
>
> > @@ -0,0 +1,86 @@
> > +require 'rake'
> > +
> > +class DatabaseAdapter
>
> Just call it Base maybe. Method name can be bigger. Also all these should
> be namespace DatabaseAdapters since they lie in that directory. In that
> case call would be DatabaseAdapters::Base.adapter_for or something
> similar.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bigbinary/boxy-cap/pull/2/files#r32737430>.
>



-- 
- Sahil

vipulnsward commented on lib/boxy-cap/database_adapter.rb:
```diff
@@ -0,0 +1,86 @@
+    def for(config, database_name)
+      case config[:adapter]
+      when /mysql/
+        require_relative 'database_adapters/mysql_database_adapter'
```
Move all require's to top of file.

sahilm replied:
The requires need to come after the `DatabaseAdapter` constant is defined because the subclasses refer to that constant.

vipulnsward replied:
The issue should go away after all classes are namespaced as mentioned in above comment.

sahilm commented:
* I have renamed `DatabaseAdapter` to `DatabaseAdapters::Base`.

* I have moved everything into the `DatabaseAdapters` module.

* I have bunched all the requires after `Base` is defined rather than in the `case` statement. Cannot figure out any way out of the lazy require though - short of putting in `autoloads`. I think this is good enough.

What do you think?

sahilm commented:
Of course I've tested this *again* with pg, mysql and sql server.
